File: frontend/src/lib/types/tasks.ts
/**
 * Tipos da carga do Hub de Tarefas em modo lista (GET /api/tarefas).
 *
 * Espelham `_serialize_task_hub_context` / `_serialize_hub_group`
 * (routes/api/tasks.py) e o serializer `serialize_task_card`
 * (routes/api/serializers.py) — a MESMA fonte de verdade da rota Jinja
 * `/tarefas` (templates/tasks/hub.html), reusada via `build_task_hub_context`
 * e respeitando o `orgao_scope` server-side.
 *
 * Task e TaskItem são a MESMA tabela: existe um único tipo `TaskCard` com os
 * CAMPOS REAIS (`descricao`, `is_archived`, `archived_at`) — sem aliases
 * legados. Os agrupamentos por projeto/etapa chegam JÁ calculados no backend;
 * o cliente NÃO recalcula a hierarquia.
 */

import type { OrgaoOption } from './pendentes';

/** Modo de listagem (`?modo=`): janela visível das tarefas do hub. */
export type TaskHubModo = 'ativas' | 'arquivadas' | 'finalizadas';

/**
 * Tarefa serializada para o card de lista (serialize_task_card + extras de
 * hub). Datas em ISO 8601 (ou `null`). `etapa_titulo` e `is_first_of_stage`
 * vêm do agrupador para que o cliente monte subgrupos por etapa sem recalcular.
 */
export interface TaskCard {
	id: number;
	descricao: string;
	status: string;
	responsavel: string | null;
	prioridade: string | null;
	tipo_pedido: string | null;
	ordem: number | null;
	project_id: number | null;
	etapa_id: number | null;
	created_by_id: number | null;
	created_at: string | null; // ISO 8601
	is_archived: boolean;
	archived_at: string | null; // ISO 8601
	/** Responsáveis múltiplos (avatares de iniciais). Substitui `responsavel`. */
	assignees: TaskAssignee[];
	/** Título da etapa à qual a tarefa pertence (ou `null` para "sem etapa"). */
	etapa_titulo: string | null;
	/**
	 * Código exibível da etapa (ex.: "42.1" = projeto.índice-visível), derivado no
	 * backend (`hub_stage_display_id`). `null` para o bucket "sem etapa".
	 */
	etapa_display_id: string | null;
	/** True quando esta tarefa abre um novo bloco de etapa no grupo. */
	is_first_of_stage: boolean;
	/** Nº de comentários (indicador na linha da lista). */
	comments_count: number;
	/** Nº de anexos (indicador na linha da lista). */
	anexos_count: number;
}

/** Grupo de tarefas de um projeto (serialize_hub_group). */
export interface TaskHubGroup {
	key: string;
	project_id: number | null;
	project_value: string;
	project_titulo: string;
	project_orgao_sigla: string | null;
	tasks: TaskCard[];
}

/** Opção de projeto para o filtro (espelha `project_options` do hub). */
export interface TaskProjectOption {
	value: string;
	label: string;
	orgao_sigla: string;
}

/** Filtros selecionados resolvidos pelo backend (`filters`). */
export interface TaskHubFilters {
	project: string;
	prioridade: string;
	tipo: string;
	status: string;
	responsavel: string;
	/** Busca livre aplicada (descrição da tarefa ou título do projeto). */
	search: string;
	/** Órgão efetivamente aplicado pelo `orgao_scope` (id sanitizado). */
	selected_orgao: string | number | null;
}

/** Metadados de paginação por GRUPO de projeto (`pagination`). */
export interface TaskHubPagination {
	page: number;
	per_page: number | null;
	total_pages: number;
	total_groups: number;
}

/** Carga completa de GET /api/tarefas (já desempacotada do envelope). */
export interface TaskHubData {
	groups: TaskHubGroup[];
	project_options: TaskProjectOption[];
	filters: TaskHubFilters;
	/** True quando a visão inclui tarefas arquivadas (`modo=arquivadas`). */
	include_archived: boolean;
	/** Total de tarefas em TODOS os grupos (pré-paginação). */
	total_items: number;
	pagination: TaskHubPagination;
	/**
	 * Opções de órgão para o filtro (mesma fonte de Pendentes:
	 * `get_user_orgao_options` + `serialize_orgao_option`; `value` = id).
	 */
	orgaos_options: OrgaoOption[];
}

/** Filtros aceitos pelo endpoint (espelham os query params de /api/tarefas). */
export interface TaskHubQuery {
	project?: string;
	prioridade?: string;
	tipo?: string;
	status?: string;
	responsavel?: string;
	/** Busca livre por descrição da tarefa ou título do projeto. */
	search?: string;
	orgao?: string | number | null;
	modo?: TaskHubModo;
	/** Página (1-based) da lista paginada por grupo de projeto. */
	page?: number;
}

/**
 * Filtros do corpo de `POST /api/tarefas/arquivar-finalizadas` (espelham
 * `_read_task_filter_values`). Mesmos nomes dos query params do hub.
 */
export interface TaskHubFilterValues {
	project?: string;
	orgao?: string | number | null;
	prioridade?: string;
	tipo?: string;
	status?: string;
	responsavel?: string;
}

/**
 * Body de `POST /api/tarefas` (criar tarefa via composer/quick-add). Espelha o
 * payload aceito por `api_tarefa_criar` (`project_id`/`etapa_id` como alias).
 */
export interface CreateTarefaInput {
	project?: string | number | null;
	project_id?: string | number | null;
	etapa?: string | number | null;
	etapa_id?: string | number | null;
	descricao: string;
	status?: string;
	responsavel?: string | null;
	/** Responsáveis múltiplos a atribuir na criação (notifica cada um). */
	assignee_ids?: number[];
	prioridade?: string | null;
	tipo_pedido?: string | null;
}

/**
 * Card devolvido por `POST /api/tarefas` (= `serialize_task_card` com
 * `with_context_labels` + `with_manage_permissions`).
 */
export interface CreatedTaskCard extends TaskCard {
	etapa_descricao: string | null;
	project_titulo: string | null;
	project_orgao_sigla: string | null;
	comments_count: number;
	anexos_count: number;
	permissions: {
		can_delete: boolean;
		can_finalize: boolean;
		is_author: boolean;
	};
}

/** Resposta de `POST /api/tarefas`. */
export interface CreateTarefaResult {
	task: CreatedTaskCard;
}

/** Resposta de `POST /api/tarefas/<id>/excluir`. */
export interface DeleteTarefaResult {
	item_id: number;
	message: string;
}

/** Resposta de `POST /api/tarefas/<id>/mover-etapa`. */
export interface MoverEtapaResult {
	task_id: number;
	etapa_id: number | null;
	previous_etapa_id: number | null;
	/** Presente quando a etapa de destino está concluída. */
	warning?: string;
}

/** Resposta de `POST /api/tarefas/arquivar-finalizadas`. */
export interface ArchiveFinalizadasResult {
	archived_count: number;
	archived_task_ids: string[];
	message: string;
}

/**
 * Responsável de uma tarefa (avatar de iniciais, sem foto). Mesma forma nos
 * candidatos do picker e em `TaskCard.assignees` — o backend serializa via
 * `serialize_assignee`.
 */
export interface TaskAssignee {
	id: number;
	name: string;
	/** Iniciais para o avatar (ex.: "AJ"). */
	initials: string;
	/** Linha secundária no dropdown (órgão ou @username). */
	subtitle?: string;
}

/** Sugestão de responsável do picker do hub (`GET .../sugestoes-responsavel`). */
export interface HubResponsavelSuggestion {
	id: number;
	name: string;
	initials: string;
	subtitle?: string;
}

File: routes/api/tasks.py
"""Endpoint JSON do Hub de Tarefas (modo lista) consumido pela SPA SvelteKit.

``GET /api/tarefas`` devolve, no envelope canônico, as tarefas do hub agrupadas
por projeto — a MESMA fonte de verdade da rota Jinja ``/tarefas``, reaproveitada
via ``build_task_hub_context`` (em ``routes/tasks/hub.py``). Respeita o escopo de
órgão server-side (``orgao_scope``); filtro de órgão inválido => 422 (em vez do
redirect 302 do Jinja).

Suporta os mesmos filtros do hub em modo lista (projeto, órgão, prioridade,
tipo, responsável, status) e um parâmetro ``?modo=`` para alternar entre as três
visões da topnav:

    - ``ativas`` (default) — tarefas não arquivadas;
    - ``arquivadas``       — tarefas arquivadas;
    - ``finalizadas``      — tarefas finalizadas (não arquivadas), espelhando o
      atalho ``/tarefas/finalizadas`` (que no Jinja redireciona para arquivadas;
      aqui filtramos por ``status == "finalizada"`` para entregar a lista direto).

Anexa ao ``main_bp`` ÚNICO (``routes/blueprint.py``); NÃO cria blueprint novo.
"""

from __future__ import annotations

from typing import Any

from flask import Response, g, request

from ..blueprint import main_bp
from ..orgao_scope import (
    get_user_orgao_options,
    sanitize_orgao_filter_for_current_user,
)
from ..tasks.hub import HUB_GROUPS_PER_PAGE, build_task_hub_context
from ..tasks.queries import _read_task_filter_values
from .envelope import fail, ok
from .negotiation import api_login_required
from .serializers import serialize_orgao_option, serialize_task_card

#: Modos de listagem aceitos por ``?modo=`` e como cada um mapeia para os
#: parâmetros de ``build_task_hub_context``. ``finalizadas`` reusa o conjunto
#: ativo filtrado por status, espelhando o atalho Jinja sem duplicar query.
_HUB_LIST_MODES = {"ativas", "arquivadas", "finalizadas"}


def _serialize_hub_group(group: dict[str, Any]) -> dict[str, Any]:
    """Serializa um grupo de projeto do hub em payload JSON-safe.

    Reusa ``serialize_task_card`` para cada tarefa do grupo (campos reais; sem
    segredos) e expõe os metadados do projeto já calculados no agrupamento
    (``_group_hub_tasks_by_project``). Anexa ``etapa_id``/``is_first_of_stage``
    a partir dos atributos ``hub_*`` definidos pelo agrupador, para que o cliente
    monte os subgrupos por etapa sem recalcular a hierarquia.

    Args:
        group: Item de ``context["groups"]`` — dict com ``project_*`` e ``tasks``.

    Returns:
        ``dict`` JSON-safe com os dados do projeto e a lista de tarefas.
    """
    tasks = []
    for task in group["tasks"]:
        card = serialize_task_card(task)
        card["etapa_titulo"] = getattr(task, "hub_stage_titulo", None)
        card["etapa_display_id"] = getattr(task, "hub_stage_display_id", None) or None
        card["is_first_of_stage"] = bool(getattr(task, "hub_is_first_of_stage", False))
        tasks.append(card)
    return {
        "key": group["key"],
        "project_id": group["project_id"],
        "project_value": group["project_value"],
        "project_titulo": group["project_titulo"],
        "project_orgao_sigla": group.get("project_orgao_sigla"),
        "tasks": tasks,
    }


def _serialize_task_hub_context(context: dict[str, Any]) -> dict[str, Any]:
    """Converte o contexto do Hub de Tarefas em payload JSON-safe.

    Serializa cada grupo de projeto via ``_serialize_hub_group`` e repassa as
    opções de projeto e os filtros selecionados já resolvidos no backend
    (``build_task_hub_context``), sem recalcular nada no cliente.

    Args:
        context: Saída de ``build_task_hub_context``.

    Returns:
        ``dict`` JSON-safe com ``groups``, ``project_options``, ``filters`` e
        ``total_items``.
    """
    return {
        "groups": [_serialize_hub_group(group) for group in context["groups"]],
        "project_options": context["project_options"],
        "filters": {
            "project": context["selected_project"],
            "prioridade": context["selected_prioridade"],
            "tipo": context["selected_tipo"],
            "status": context["selected_status"],
            "responsavel": context["selected_responsavel"],
            "search": context["selected_search"],
            "selected_orgao": context["selected_orgao"],
        },
        "include_archived": context["include_archived"],
        "total_items": context["total_items"],
        "pagination": context["pagination"],
        # Opções de órgão para o filtro local — MESMA fonte de Pendentes
        # (get_user_orgao_options, escopo server-side), com value = ID de
        # OrgaoUnidade. A SPA antes derivava SIGLAS de project_options e o
        # sanitizador (que espera id) rejeitava qualquer escolha com 422.
        "orgaos_options": [
            serialize_orgao_option(node) for node in get_user_orgao_options(g.user)
        ],
    }


def _resolve_hub_list_mode(raw_mode: str) -> tuple[bool, str, str]:
    """Resolve o ``?modo=`` em (include_archived, status_filter, modo_normalizado).

    Args:
        raw_mode: Valor cru de ``?modo=`` (vazio => "ativas").

    Returns:
        Tupla ``(include_archived, status_override, modo)``:
            - ``ativas``     -> ``(False, "", "ativas")``;
            - ``arquivadas`` -> ``(True, "", "arquivadas")``;
            - ``finalizadas``-> ``(False, "finalizada", "finalizadas")``.
        Modo desconhecido cai em "ativas".
    """
    modo = (raw_mode or "ativas").strip().lower()
    if modo not in _HUB_LIST_MODES:
        modo = "ativas"
    if modo == "arquivadas":
        return True, "", modo
    if modo == "finalizadas":
        return False, "finalizada", modo
    return False, "", modo


@main_bp.route("/api/tarefas", methods=["GET"])
@api_login_required
def api_tarefas() -> Response | tuple[Response, int]:
    """Retorna o Hub de Tarefas (modo lista) no envelope canônico para a SPA.

    Reaproveita ``build_task_hub_context`` (a mesma fonte usada pela rota Jinja
    ``/tarefas``) e respeita o escopo de órgão server-side. O filtro ``?orgao=``
    é sanitizado para o usuário corrente; um valor inválido (fora do escopo)
    resulta em 422 JSON (``validation``) em vez do redirect 302 do fluxo Jinja.
    Os parâmetros ``?project=``, ``?prioridade=``, ``?tipo=`` e ``?responsavel=``
    espelham os filtros do hub; ``?search=`` busca por substring na descrição da
    tarefa ou no título do projeto; ``?modo=`` alterna entre ``ativas`` (default),
    ``arquivadas`` e ``finalizadas``. ``?page=`` pagina por GRUPO de projeto
    (``HUB_GROUPS_PER_PAGE`` grupos/página; metadados em ``pagination``).

    Returns:
        Envelope ``{"ok": true, "data": {...}}`` com HTTP 200; ou
        ``fail(..., 422, "validation")`` quando o filtro de órgão é inválido.
        ``api_login_required`` devolve 401 JSON quando não há sessão.
    """
    filter_values = _read_task_filter_values(request.args)
    selected_orgao_id, invalid_orgao_filter = sanitize_orgao_filter_for_current_user(
        filter_values["orgao_filter"]
    )
    if invalid_orgao_filter:
        return fail(
            "Filtro de órgão inválido para o usuário.",
            status=422,
            code="validation",
        )

    include_archived, status_override, _modo = _resolve_hub_list_mode(
        request.args.get("modo", "")
    )
    status_filter = status_override or filter_values["status_filter"]

    context = build_task_hub_context(
        project_filter=filter_values["project_filter"],
        prioridade_filter=filter_values["prioridade_filter"],
        tipo_filter=filter_values["tipo_filter"],
        status_filter=status_filter,
        responsavel_filter=filter_values["responsavel_filter"],
        search_filter=filter_values["search_filter"],
        selected_orgao_id=selected_orgao_id,
        apenas_orgao=filter_values["apenas_orgao_filter"],
        include_archived=include_archived,
        page=request.args.get("page", 1, type=int),
        per_page=HUB_GROUPS_PER_PAGE,
    )
    return ok(_serialize_task_hub_context(context))
